launcher/net/Download.cpp

```diff
@@ -48,12 +51,106 @@
+
 namespace Net {
 
+QString truncateUrlHumanFriendly(QUrl &url, int max_len, bool hard_limit = false)
+{   
+    auto display_options = QUrl::RemoveUserInfo | QUrl::RemoveFragment | QUrl::NormalizePathSegments;
+    auto str_url = url.toDisplayString(display_options);
+    if (str_url.length() <= max_len)
+        return str_url;
+
+    /* this is a PCRE regular expression that splits a URL (given by the display rules above) into 5 capture groups
+     * the scheme (ie https://) is group 1
+     * the host (with trailing /) is group 2
+     * the first part of the path (with trailing /) is group 3
+     * the last part of the path (with leading /) is group 5
+     * the remainder of the URL is in the .* and in group 4
+     *
+     * See: https://regex101.com/r/inHkek/1
+     * for an interactive breakdown
+     */ 
+    QRegularExpression re(R"(^([\w]+:\/\/)([\w._-]+\/)([\w._-]+\/)(.*)(\/[^]+[^]+)$)");
+    
+    auto url_compact = QString(str_url);
+    url_compact.replace(re, "\\1\\2\\3...\\5");
+    if (url_compact.length() >= max_len) {
+        url_compact = QString(str_url);
+        url_compact.replace(re, "\\1\\2...\\5");
+    }
+
+
+    if ((url_compact.length() >= max_len) && hard_limit) {
+        auto to_remove = url_compact.length() - max_len + 3;
+        url_compact.remove(url_compact.length() - to_remove - 1, to_remove);
+        url_compact.append("...");
+    }
+
+    return url_compact;
+
+}
+
+QString humanReadableDuration(double duration, int precision = 0) {
+
+    using days = std::chrono::duration<int, std::ratio<86400>>;
+
+    QString outStr;
+    QTextStream os(&outStr);
+
+    auto std_duration = std::chrono::duration<double>(duration);
+    auto d = std::chrono::duration_cast<days>(std_duration);
+    std_duration -= d;
+    auto h = std::chrono::duration_cast<std::chrono::hours>(std_duration);
+    std_duration -= h;
+    auto m = std::chrono::duration_cast<std::chrono::minutes>(std_duration);
+    std_duration -= m;
+    auto s = std::chrono::duration_cast<std::chrono::seconds>(std_duration);
+    std_duration -= s;
+    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(std_duration);
+
+    auto dc = d.count();
+    auto hc = h.count();
+    auto mc = m.count();
+    auto sc = s.count();
+    auto msc = ms.count();
+
+    if (dc) {
+        os << dc << "days";
+    }
+    if (hc) {
+        if (dc)
+            os << " ";
+        os << qSetFieldWidth(2) << hc << "h";
+    }
+    if (mc) {
+        if (dc || hc)
+            os << " ";
+        os << qSetFieldWidth(2) << mc << "m";
+    }
+    if (dc || hc || mc || sc) {
+        if (dc || hc || mc)
+            os << " ";
+        os << qSetFieldWidth(2) << sc << "s";
+    }
+    if ((msc && (precision > 0)) || !(dc || hc || mc || sc)) {
+        if (dc || hc || mc || sc)
+            os << " ";
+        os << qSetFieldWidth(0) << qSetRealNumberPrecision(precision) << msc << "ms";
+    }
+
+    os.flush();
+
+    return outStr;
+}
+
```

Comment:
This should probably go into `launcher/MMCTime.cpp`.

Also, having some tests for this code would be nice, but not a requirement for merging.